### PR TITLE
bpo-43976: add vendor config

### PR DESCRIPTION
**Pull request by FFY00**
Signed-off-by: Filipe Laíns <lains@riseup.net>

Python-Dev proposal:
https://gist.github.com/FFY00/625f65681fbcd7fc039dd4d727bb2c2f
https://mail.python.org/archives/list/python-dev@python.org/thread/OWETD7ZHCHX3R4IMGO3YPRC6WABVWGXH/

https://bugs.python.org/issue43976

### CODE REVIEW

**Comment by encukou** [comment]
The "Check if generated files are up to date" test is failing because of unfortunate timing when merging other PRs (GH-25315, GH-25687).
It's OK to ignore it. Sorry for the inconvenience.
To make it pass, you can merge the master branch.

**Comment by FFY00** [comment]
No worries, I haven't committed the generated files anyway. I purposedly excluded generated files from the commit to make it easier to review. I still have to add documentation to the commit, and then I will push a commit updating the generated files, I can rebase when I do that. The only reason I opened this PR now anyway was that I am not gonna be home this afternoon and wanted to add the missing bits from my laptop.
Before I mark it ready for review I will be sure to rebase (I usually do) :blush:

**Comment by tiran** [comment]
Please use autoconf 2.69 to regenerate configure scripts and helpers. Most core devs are on platforms that do not have autoconf 2.71 yet.

**Comment by FFY00** [comment]
The venv failure seems to be because we are hitting https://github.com/pypa/pip/issues/9617.

**Comment by merwok** [comment]
I think this needs a mailing-list or forum discussion.

I would have notified the people in https://github.com/pypa/pip/issues/9617 but that is locked, maybe @uranusjr or @pradyunsg can ping them.  (But as all design discussions, please use a proper python forum and not a PR)

**Comment by FFY00** [comment]
I have opened https://discuss.python.org/t/mechanism-for-distributors-to-add-site-install-schemes-to-python-installations/8467.

A note about the PR. Currently, the site module does not import sysconfig, but it contains two copied functions from it. As far as I understand, this is just a micro optimization, the only two sysconfig functions it needs are pretty simple and do not depend on other sysconfig code or other modules.
This PR, however, needs a significant part of sysconfig, so it imports it.

I did some bechmarks to see if we would notice any performance difference, but the results are pretty negligible.

With the PR:
```
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 167 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 174 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 166 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 176 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 184 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 176 nsec per loop
```

In `master`:
```
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 176 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 179 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 171 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 176 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 177 nsec per loop
$ ./python -m timeit 'import site'
2000000 loops, best of 5: 176 nsec per loop
```

If this PR goes in, I will open a new PR removing the two sysconfig functions that the site module vendors.

**Comment by merwok** [comment]
Can you build python from scratch if sysconfig imports site?
What about cross-compilation? (which has unclear support status)

**Comment by pradyunsg** [comment]
For the stuff with the pip release, the PR that needs merging is https://github.com/pypa/pip/pull/9912. Once that's merged and released, things should get fixed.

**Comment by FFY00** [comment]
I have applied #25761 to fix the tests, github should be able to merge this without issue after that PR goes in.

File: Doc/library/site.rst
Comment (pradyunsg) [inline Doc/library/site.rst]: There's a typo!

Comment (FFY00) [reply Doc/library/site.rst]: Thanks :sweat_smile:

**Comment by FFY00** [comment]
@uranusjr can you take a look at the last commit (the `get_preferred_schemes` one)?

File: Lib/test/test_site.py
Comment (jaraco) [inline Lib/test/test_site.py]: Would it make sense to have a separate test rather than mutate an existing test?

File: Lib/test/test_sysconfig.py
Comment (jaraco) [inline Lib/test/test_sysconfig.py]: Would it make sense to have two tests, one which patches sys.path and another that doesn't?

**Review by jaraco — CHANGES_REQUESTED** [review]
This looks good to me and a good start for providing hooks for vendor configurability.

I'd like to see an answer to @merwok's questions:

> Can you build python from scratch if sysconfig imports site?
> What about cross-compilation? (which has unclear support status)

I'm not familiar with building from scratch, but I agree, it would be important to address these cases. Another way to think about it - is this change backward-compatible? Can this change provide a graceful degradation such that if no vendor file is supplied, the behavior is all-but-guaranteed to be the same as on prior versions (if no `--vendor-config` is supplied)? If the change is "safe" in that respect (compatibility), the feature can be refined through the betas.

@merwok Do you know how one could test for building Python from scratch (to prove compatibility or demonstrate a failure)?

Also, the docs builds are failing - so those checks will need to be fixed before merging.

Those things addressed, I'm ready to say this change is suitable for merge, acknowledging that the short time frame prior to the feature freeze means this change may have a fair likelihood of being reverted (with all the consternation that brings and leaving no opportunity to restore prior to Python 3.11).

File: Lib/test/test_site.py
Comment (FFY00) [reply Lib/test/test_site.py]: Sure, we still need the test as-is, but we can also keep the original one.

File: Lib/test/test_sysconfig.py
Comment (FFY00) [reply Lib/test/test_sysconfig.py]: Ditto.

**Comment by FFY00** [comment]
I dug into the doc issue, it seems to me that pip is detecting some mismatch in sysconfig and then deciing to use `{sysconfig.get_path('scripts')}/python` instead of `sys.executable` as the interpreter for the entrypoint scripts.

cc @pradyunsg @uranusjr

**Comment by FFY00** [comment]
> I'd like to see an answer to @merwok's questions:
> 
> > Can you build python from scratch if sysconfig imports site?
> > What about cross-compilation? (which has unclear support status)
> 
> I'm not familiar with building from scratch, but I agree, it would be important to address these cases. Another way to think about it - is this change backward-compatible? Can this change provide a graceful degradation such that if no vendor file is supplied, the behavior is all-but-guaranteed to be the same as on prior versions (if no `--vendor-config` is supplied)? If the change is "safe" in that respect (compatibility), the feature can be refined through the betas.
>
> @merwok Do you know how one could test for building Python from scratch (to prove compatibility or demonstrate a failure)?

I don't understand, what is building "from scratch" here? CPython does not have an alternative build process for bootstrapping, which is what would generally be considered a "build from scratch". It's just "building CPython"... and it's what the CI already tests.
Well, I guess the CI doesn't really test installing on the system and checking if that works, but yes, it does work.

`site` importing `sysconfig` is not an issue, it works fine. I think this is what @merwok was worried about, right? But as I said above, there is nothing special "build from scratch" process, like there is in for eg. GCC, so this is already tested in the CI.

Anyway, I have built two arch packages, pretty much the same as what I would actually deploy, with and without `--with-vendor-config` and everything works as expected.

I haven't tried cross-compiling, but I can't see how that would have any impact here, we are not touching any related files, it's quite literally just some generic Python machinery.

@merwok could you clarify your question(s)? And let me know if my comments here answered what you were looking for.

> Also, the docs builds are failing - so those checks will need to be fixed before merging.

Fixed now, see the comment above.

**Comment by merwok** [comment]
> It's just "building CPython"

Yes this is what I meant.  Some modules have a notes that they copy or hard-code some things because they are needed when building CPython (as opposed to using an installed CPython).

> and it's what the CI already tests

Then the change is compatible!

**Review by jaraco — APPROVED** [review]

**Comment by tiran** [comment]
Hi, I would prefer to get feedback from our resident Debian/Ubuntu and Fedora/RHEL experts before merging the PR.

**Comment by FFY00** [comment]
> > It's just "building CPython"
> 
> Yes this is what I meant. Some modules have a notes that they copy or hard-code some things because they are needed when building CPython (as opposed to using an installed CPython).
> 
> > and it's what the CI already tests
> 
> Then the change should be compatible!

Thanks for clarifying!

> > I haven't tried cross-compiling, but I can't see how that would have any impact here
> 
> Not foreseeing a problem is not a strong position compared to positive confirmation that what’s needed still works. I can’t give more details beside raising the point as I don’t cross-compile myself, but hopefully the people handling redistribution will weigh in in the discussion thread and validate these changes.

Right, I don't have a easy way to test it at the moment, that's why I didn't do it.

**Comment by FFY00** [comment]
> Hi, I would prefer to get feedback from our resident Debian/Ubuntu and Fedora/RHEL experts before merging the PR.

@encukou and @doko42? Am I missing anyone?

**Comment by pradyunsg** [comment]
@stefanor and @hroncok, I think?

**Comment by hroncok** [comment]
I'd like to have @frenzymadness look at this as well.

File: Lib/sysconfig.py
Comment (hroncok) [inline Lib/sysconfig.py]: If this check was not here, we could use this instead of our patch entirely. I'm not sure what is the benefit of the check.

**Comment by pradyunsg** [comment]
Arch Linux: @felixonmars (maintainer), @eli-schwartz (community contributor)

**Comment by hroncok** [comment]
Another thing I am not sure about is the performance impact on importing sysconfig from site. I know @vstinner recently worked hard to reduce the number of modules imported by default.

File: Lib/sysconfig.py
Comment (FFY00) [reply Lib/sysconfig.py]: That breaks several assumptions in other parts of the code, so I chose to not go into that right now. We could remove it in the future if it there is a valid use case, though I can't really think of anything strong enough to warrant that.

**Comment by pradyunsg** [comment]
Well, that's probably enough folks being @-mentioned here now. ;)

**Comment by FFY00** [comment]
Just to clarify and help all the mentioned people understand the intent, this allows Python distributors to solve the issue of sharing `site-packages` with tools like pip. In the config, you can add extra install schemes, and you can make it so that they are activated in the site module initialization.

Example:
```python
EXTRA_INSTALL_SCHEMES = {
    'vendor': {
        'stdlib': '{installed_base}/{platlibdir}/python{py_version_short}',
        'platstdlib': '{platbase}/{platlibdir}/python{py_version_short}',
        'purelib': '{base}/lib/python{py_version_short}/vendor-packages',
        'platlib': '{platbase}/{platlibdir}/python{py_version_short}/vendor-packages',
        'include':
            '{installed_base}/include/python{py_version_short}{abiflags}',
        'platinclude':
            '{installed_platbase}/include/python{py_version_short}{abiflags}',
        'scripts': '{base}/bin',
        'data': '{base}',
    },
}

EXTRA_SITE_INSTALL_SCHEMES = [
    'vendor',
]
```

This will register a new scheme, `vendor`, which will use a `vendor-packages` folder, and it will be activated during the site module initialization (`vendor-packages` will be in the import path).
Using this, Python distributors can put their schemes in this directory, solving the issue I described above.
It also makes it possibly to create entrypoint scripts that only use these packages, by making them start the interpreter without the site module initialization (`-S`) and then calling `site.addsitedir(sysconfig.get_path('purelib', 'vendor'))`, preventing for eg. `pip install blah` from breaking distro-provided scripts, which results in a better UX -- less systems breaking. This last one is just a nice thing that will not be possible.

**Comment by FFY00** [comment]
@tiran I don't think we will get all the feedback you requested before the feature freeze, and we were hoping to get it in for 3.10 due to the distutils deprecation. This issue has already been discussed extensively, and the approach has also been discussed in some places.
Would it be possible to get the PR in Monday, in time for the next beta, and if any issue arises yank it from 3.10? We should also be able to fix minor issues in the next betas if needed.

**Comment by hroncok** [comment]
Since `distutils.command.install` does not read/respect the result of `get_preferred_schemes()`, we (Fedora) still need to patch the default `posix_prefix` (`unix_prefix` in distutils nomenclature) to do the logic for us.

Currently, we patch `distutils` directly, but we need to avoid that, so it doesn't matter if `distutils` comes from the standard library, PyPI or `setuptools._vendor`.

The mechanism proposed here explicitly prevents us from updating the existing schemes and only allows us to add additional ones. If merged, we'd be stuck in a position where we need to select from the two options:

- patch `sysconfig._INSTALL_SCHEMES`
- use this mechanism and still patch `sysconfig._INSTALL_SCHEMES`

Since we would still need to patch `sysconfig._INSTALL_SCHEMES`, using this mechanism only adds extra complexity for us. When choosing whether to use it, we would likely avoid it because of that.

This does not block Fedora from doing what we have planned. But in the current form, it doesn't really help either.

**Comment by FFY00** [comment]
IMO `distutils.command.install` should be fixed before the deprecation, but we are essentially in the last day of the feature freeze. I think it would be reasonable to fix `distutils.command.install` in a future beta, or postpone the distutils deprecation. @jaraco what do you think?

**Comment by tiran** [comment]
@FFY00 a release manager can grant special exception from the beta rule. This feature is relevant for downstream packagers. The deprecation of distutils is a special circumstance, too. I would prefer to delay this improvement to beta 2 instead of landing a not-quite-ready solution in beta 1.

Paging @pablogsal

File: Lib/sysconfig.py
Comment (jaraco) [reply Lib/sysconfig.py]: FFY00, are you saying that removing the check would break several assumptions in other parts of the code? Can you provide an example of such a broken assumption? Isn't "vendor doesn't have to patch sysconfig" a compelling reason to warrant it (given that's the goal of this patch)?

Comment (FFY00) [reply Lib/sysconfig.py]: > FFY00, are you saying that removing the check would break several assumptions in other parts of the code? Can you provide an example of such a broken assumption?

There are parts of the `site`, `venv` and `distutils` modules, for example, that hardcode `site-packages`. Allowing vendors to change this path would cause at the minimum a moderate amount of breakage.

> Isn't "vendor doesn't have to patch sysconfig" a compelling reason to warrant it (given that's the goal of this patch)?

It is, though, generally there are several parts of the code that can be change to solve such issues, and I think there might be easier, and probably better, ways to do it rather than allowing vendors to change the default install schemes.

So the question here should be, why would vendors need to change the default schemes? Can we solve the issue in a different way? What would make more sense, removing this check, or any of the alternatives?

Fedora's case can be solved by changing `distutils.command.install` to respect `get_preferred_schemes`, which I feel is a much better solution.
If there are any other reasons where someone might want to remove this check, we can look at them and decide what makes the most sense.

**Comment by jaraco** [comment]
> Since we would still need to patch `sysconfig._INSTALL_SCHEMES`, using this mechanism only adds extra complexity for us. When choosing whether to use it, we would likely avoid it because of that.

What's wrong with another possible approach:

- don't patch `sysconfig._INSTALL_SCHEMES`.
- Supply additional install vendor-specific install schemes, to be honored by setuptools.
- Patch stdlib distutils to also supply these vendor-specific install schemes (for compatibility until everything honors the schemes in sysconfig).

> IMO `distutils.command.install` should be fixed before the deprecation, but we are essentially in the last day of the feature freeze. I think it would be reasonable to fix `distutils.command.install` in a future beta, or postpone the distutils deprecation. @jaraco what do you think?

I was imagining that distutils would either be patched by the vendor environment or would honor the new sysconfig install schemes. I have no opinion on which approach would be better. Yes, stdlib distutils needs to work, but only for compatibility cases. The primary, long term supported cases will migrate to Setuptools and sysconfig. The key contribution here is that vendors have a clean and _supported_ mechanism to supply install schemes and other customizations for downstream packaging tools.

> The mechanism proposed here explicitly prevents us from updating the existing schemes and only allows us to add additional ones.

I believe this issue is at the crux of the design. Looking at [our notes](https://hackmd.io/r-iOdmM_TGSiYpyGKp-cFg?view#2021-04-03-Notes), we were planning on allowing "override" of install schemes. @FFY00 Can you elaborate on the issues with allowing overrides of the default install schemes?

**Comment by FFY00** [comment]
> I believe this issue is at the crux of the design. Looking at [our notes](https://hackmd.io/r-iOdmM_TGSiYpyGKp-cFg?view#2021-04-03-Notes), we were planning on allowing "override" of install schemes. @FFY00 Can you elaborate on the issues with allowing overrides of the default install schemes?

I have already mentioned it above, but essentially several parts of the code are hardcoding the path, I assume to avoid a `sysconfig` import. Also, I suspect there are a few other non-obvious cases of similar assumptions.

(from my reply above)
> There are parts of the `site`, `venv` and `distutils` modules, for example, that hardcode `site-packages`. Allowing vendors to change this path would cause at the minimum a moderate amount of breakage.

There are also users relying on this name, which would be OK to break, but I would rather prevent it if possible.
https://github.com/pypa/pip/blob/e6414d6db6db37951988f6f2b11ec530ed0b191d/setup.cfg#L94
https://github.com/pypa/virtualenv/blob/45b44c20a3b7849c97ca17ec86b3b49ead2314dd/.coveragerc#L19

**Comment by hroncok** [comment]
> Patch stdlib distutils to also supply these vendor-specific install schemes (for compatibility until everything honors the schemes in sysconfig).

If we do this and setuptools bring their own distutils, our patch is overridden when user update setuptools (e.g. via pip). That is the reason we can no longer patch distutils.

> IMO distutils.command.install should be fixed before the deprecation...

Yes, it should have been, but it is too late for that in 3.10. It is not a trivial fix.

File: Lib/site.py
Comment (vstinner) [inline Lib/site.py]: sysconfig creates tons of Python objects and is slow to import. I would prefer to avoid it at startup.

Would it be possible to move the logic into a smaller module which does exactly what you need, and reuse this module in sysconfig? See _sitebuiltins for example.

Comment (FFY00) [reply Lib/site.py]: Yes, I think this should be perfectly doable. It also stacks up another reason to implement @zooba's suggestion of making sysconfig a directory and installing the vendor config as a submodule, to prevent shadowing it.

Comment (uranusjr) [reply Lib/site.py]: > It also stacks up another reason to implement @zooba's suggestion of making sysconfig a directory and installing the vendor config as a submodule, to prevent shadowing it.

Maybe I’m misunderstanding, but are you hinting to put the smaller module in a subpackage? Since that won’t work because `import sysconfig.bare` implicitly imports `sysconfig` as well so unless you move all names out of `sysconfig` (not an option since it breaks all existing code) this will not help import time.

Comment (FFY00) [reply Lib/site.py]: Right :facepalm:

**Comment by doko42** [comment]
This was also a topic at the Python Language Summut 2021, followed by an open space at PyCon, and a follow-up with a sprint.  There is a need to have something like a vendor site-packages, but this patch comes a little bit short.

- The patch adds a new directory on sys.path, but doesn't define any semantics for this vendor site-packages
- What does python -S mean for the vendor site-packages?
- What is the preferred scheme?  At least for Debian and Ubuntu, we have three prefixes for the system CPython: /usr/lib/python3/dist-packages (which is the proposed new directory), /usr/local/lib/pythonX.Y/dist-packages (which is empty and is the location where local installations made by a site administrator should go), and site-packages within virtual envs, which is linked to /usr/lib/python3/dist-packages. /usr/local/lib/pythonX.Y/dist-packages is not exposed in virtual envs.
- Debian carries a patch (currently only for distutils.sysconfig to make these schemes known, with a runtime check whether CPython is used on the system, or in a virtual env. I'm worried that I still have to patch sysconfig, and using the new configure option. In this case I would just prefer a local patch.
- The current situation, and the current proposed patch still have the problem distinguishing about header files and libraries installed by extensions.  You want to use the vendor schema to get the location of the CPython headers and libraries, but you don't want to install into these locations, but more likely into something like /usr/local/include

I would prefer to defer this patch until we (some distro guys, some PyPA guys, some CPython developers, including Filipe) come up with a draft for a PEP.

**Comment by FFY00** [comment]
> The patch adds a new directory on sys.path, but doesn't define any semantics for this vendor site-packages

The vendor site-packages is the same as the normal site-packages, they are just one extra location, so same semantics.

> What does python -S mean for the vendor site-packages?

They get treated the same as normal site-packages, as one would expect.

> What is the preferred scheme? At least for Debian and Ubuntu, we have three prefixes for the system CPython: /usr/lib/python3/dist-packages (which is the proposed new directory), /usr/local/lib/pythonX.Y/dist-packages (which is empty and is the location where local installations made by a site administrator should go), and site-packages within virtual envs, which is linked to /usr/lib/python3/dist-packages. /usr/local/lib/pythonX.Y/dist-packages is not exposed in virtual envs.

That can be defined by the vendor in `get_preferred_scheme`, which evaluates at runtime.

> Debian carries a patch (currently only for distutils.sysconfig to make these schemes known, with a runtime check whether CPython is used on the system, or in a virtual env. I'm worried that I still have to patch sysconfig, and using the new configure option. In this case I would just prefer a local patch.

The current patch does not disable the vendor schemes in a virtual environment, they are technically there but no one uses them, and they not point to the system path unless the vendor explicitly uses the full path. This would be the recommended configuration:

```py
EXTRA_SITE_INSTALL_SCHEMES = {
    'vendor': {
        'stdlib': '{installed_base}/{platlibdir}/python{py_version_short}',
        'platstdlib': '{platbase}/{platlibdir}/python{py_version_short}',
        'purelib': '{base}/lib/python{py_version_short}/vendor-packages',
        'platlib': '{platbase}/{platlibdir}/python{py_version_short}/vendor-packages',
        'include':
            '{installed_base}/include/python{py_version_short}{abiflags}',
        'platinclude':
            '{installed_platbase}/include/python{py_version_short}{abiflags}',
        'scripts': '{base}/bin',
        'data': '{base}',
    },
}
```

Which will use the virtual environment base, and will not point to the system.

> The current situation, and the current proposed patch still have the problem distinguishing about header files and libraries installed by extensions. You want to use the vendor schema to get the location of the CPython headers and libraries, but you don't want to install into these locations, but more likely into something like /usr/local/include

Only if you explicitly choose to point `include`/`platinclude` to a different path, with the config above we will get the expected location, `/usr/include/python3.10`.

The problem here seems to be that Debian seems to taking a liberal opinionated of the FHS and wanting to hijack a different prefix than the software was compiled for.

https://refspecs.linuxfoundation.org/FHS_3.0/fhs.html#usrlocalLocalHierarchy

The FHS says that `/usr/local` should be used by system administrator when installing software locally. Debian interprets this as mandating that user packages go to `/usr/local`, but still having those packages affect the operation of the Python in `/usr`. Such package *are a part* of the `/usr` installation, they are components that affect its operation.

This is the reason Debian had to come up with `dist-packages` in the first place, to stop conflicting with local Python installations because they want to use `/usr/local`. The software was simply not designed for this, so such workarounds are required, but at the end of the day this just end up causing years of pain to Python users and developers.

What the FHS refers to here is that software, like a new Python installation, installed locally by a system administrator should go into `/usr/local`. It does not mean that components of the certain installation should go into `/usr/local` if they happen to be managed by the local system administrator. Those would be *customizations* of the current installation, *not installations themselves*. The software as far the FHS is concern is Python itself, only.

I am more than happy to raise this issue on the FHS mailing list, I am also happy to raise this with the Debian technical board or equivalent.

If you drop the requirement to be able to install a Python interpreter managed scheme outside its prefix, there are no issues.

> I would prefer to defer this patch until we (some distro guys, some PyPA guys, some CPython developers, including Filipe) come up with a draft for a PEP.

If that's the case, I would then request the distutils deprecation to be postponed to 3.11.

**Comment by doko42** [comment]
> > The current situation, and the current proposed patch still have the problem distinguishing about header files and libraries installed by extensions. You want to use the vendor schema to get the location of the CPython headers and libraries, but you don't want to install into these locations, but more likely into something like /usr/local/include
> 
> Only if you explicitly choose to point `include`/`platinclude` to a different path, with the config above we will get the expected location, `/usr/include/python3.10`.
> 
> The problem here seems to be that Debian seems to taking a liberal opinionated of the FHS and wanting to hijack a different prefix than the software was compiled for.

No, this is unrelated. either you have the same location in both schemas leading to some conflicts, or you have on to get the headers from and one to write the theaders to.

**Comment by FFY00** [comment]
Packages don't install to `include`/`platinclude`, they are currently shared between all schemes.

**Comment by FFY00** [comment]
> No, this is unrelated. either you have the same location in both schemas leading to some conflicts, or you have on to get the headers from and one to write the theaders to.

You are in control of the schemes, you can change the `include`/`platinclude` paths to avoid conflicts.

I agree that that might not be the nicest solution, but the CPython upstream is in control of the default schemes, so it would be trivial to change the scheme definitions to use subdirectories, similarly to `purelib`/`platlib`.

**Review by youtacrsbyvars — APPROVED** [review]

File: Lib/site.py
Comment (vstinner) [reply Lib/site.py]: you can add a `_my_bootstrap_module` module and import it in `sysconfig` and `site` modules. It doesn't change the API, only the implementation, and Python startup time: https://pythondev.readthedocs.io/startup_time.html

File: Doc/Makefile
Comment (tiran) [inline Doc/Makefile]: This looks like an unrelated change and breaks building of documentation w/o virtual envs

Comment (FFY00) [reply Doc/Makefile]: It is, see [`aa36fb5` (#25718)](https://github.com/python/cpython/pull/25718/commits/aa36fb5850e366ab7d0225faaa09f8080cb6d251), I don't think it's required anymore.

File: Doc/library/site.rst
Comment (tiran) [inline Doc/library/site.rst]: It's a new feature, so 3.11.
```suggestion
.. versionchanged:: 3.11
```

File: aclocal.m4
Comment (tiran) [inline aclocal.m4]: Please use https://quay.io/repository/tiran/cpython_autoconf to regenerate the autoconf files.

**Review by tiran — CHANGES_REQUESTED** [review]

**Comment by FFY00** [comment]
I ran the `python_startup` benchmark:

```
2021-10-19_13-44-main-09c04e7f0d26.json.gz
==========================================

Performance version: 1.0.2
Report on Linux-5.14.9-hardened1-1-hardened-x86_64-with-glibc2.33
Number of logical CPUs: 80
Start date: 2021-10-19 17:11:21.690235
End date: 2021-10-19 17:12:13.904349

2021-10-19_17-52-master-5cda6f0d120e.json.gz
============================================

Performance version: 1.0.2
Report on Linux-5.14.9-hardened1-1-hardened-x86_64-with-glibc2.33
Number of logical CPUs: 80
Start date: 2021-10-19 19:17:32.214856
End date: 2021-10-19 19:18:28.265266

### python_startup ###
Mean +- std dev: 21.3 ms +- 0.1 ms -> 23.1 ms +- 0.1 ms: 1.09x slower
Significant (t=-170.97)
```

**Comment by FFY00** [comment]
> please avoid history rewriting and force pushes for CPython, they give a poor experience for reviewsers (this is a devguide guideline)

Sorry abou that, I originally forced pushed because of the rebase, and then I made some very small fixups that for mistakes I made.

> could you explain the sysconfig/_sysconfig split in module docstrings?

Will do.

> could an alternative technique be used (freezing the module) to avoid splitting and losing history because of git’s poor support for renaming?

freezing could help, but most of the problematic attributes need to be dynamically evaluated when the user executes the code, so we wouldn't get much improvement. I am currently working on lazy loading the most of the problematic attributes, which I believe should be more successful, but it requires a big refactor of the module, which would not be as much different as renamig, and we would still benefit from the renaming anyway.

> not a big fan of adding `_vendor.config` submodule; `concurrent.futures` showed us that too much nesting can prove unnecessary

Putting the config in a submodule was proposed by @zooba IIRC, to make it more difficult for users to shadow it by mistake.

**Comment by FFY00** [comment]
I am closing this now. After 2 separate attempts at getting this merged, no progress was made. #29041 made a similar change to the interpreter initialization and did not have near this much push back, which, to be clear, I think that was very good progress, but so is this IMO, it solves a very real problem. I am really unsure about what I could have done differently to have this progress further.

I'd welcome any feedback, and other alternative approaches to solve this problem, so far I don't think anyone made any other serious proposal.

**Comment by merwok** [comment]
This comment https://github.com/python/cpython/pull/25718#issuecomment-843135164 expressed that a broad discussion was needed and its outcome encoded in a PEP, which makes sense to me.  There’s a lot of history of CPython devs doing changes and OS maintainers adding their patches without much communication (also seen recently with the surprise and resistance to backend-independent packaging even though it’s been years in the making and communicated in multiple places); it would be good to see all these people agree on the needs, problems and solutions.

**Comment by FFY00** [comment]
Thank you for the feedback, I realized that and ended up witting https://gist.github.com/FFY00/625f65681fbcd7fc039dd4d727bb2c2f, but was hoping it did not have to make it to a full-blown PEP. I have gotten feedback from a couple people, and for all of the ones that could benefit from this mechanism, this seemed like a reasonable approach. Unfortunately, I did not get any feedback from Matthias, who was the main blocker on this PR so I am not sure what else I can do besides submitting it as a PEP, but I am not sure we will get the feedback you were hopping from everyone/

**Pull request closed without merging**
